packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart

```diff
@@ -0,0 +1,407 @@
+      assert(constraints == this.constraints);
+      allowAdditionsFor(index, () {
+        _childManager.createChild(index, after: null);
+      });
+    });
+    if (indexOf(firstChild) == index) {
+      assert(parentUsesSize);
+      firstChild.layout(childConstraints, parentUsesSize: parentUsesSize);
+      SliverMultiBoxAdaptorParentData firstChildParentData = firstChild.parentData;
+      firstChildParentData.scrollOffset = endScrollOffset - paintExtentOf(firstChild);
+      return firstChild;
+    }
+    return null;
+  }
+
+  /// Called during layout to create, add, and layout the child after
+  /// the given child.
+  ///
+  /// Calls [createChild] to actually create and add the child.
+  ///
+  /// Returns the new child. It is the responsibility of the caller to configure
+  /// the child's scroll offset.
+  ///
+  /// Children after the `after` child may be removed in the process. Only the
+  /// new child may be added.
+  @protected
+  RenderBox insertAndLayoutChild(BoxConstraints childConstraints, {
+    @required RenderBox after,
+    bool parentUsesSize: false,
+  }) {
+    assert(_currentlyUpdatingChildIndex == null);
+    assert(after != null);
+    final int index = indexOf(after) + 1;
+    invokeLayoutCallback<SliverConstraints>((SliverConstraints constraints) {
+      assert(constraints == this.constraints);
+      allowAdditionsFor(index, () {
+        _childManager.createChild(index, after: after);
+      });
+    });
+    final RenderBox child = childAfter(after);
+    if (child != null && indexOf(child) == index) {
+      assert(indexOf(child) == index);
+      child.layout(childConstraints, parentUsesSize: parentUsesSize);
+      return child;
+    }
+    return null;
+  }
+
+  /// Called after layout with the number of children that can be garbage
+  /// collected at the head and tail of the child list.
+  @protected
+  void collectGarbage(int leadingGarbage, int trailingGarbage) {
+    assert(_currentlyUpdatingChildIndex == null);
+    assert(childCount >= leadingGarbage + trailingGarbage);
+    invokeLayoutCallback<SliverConstraints>((SliverConstraints constraints) {
+      while (leadingGarbage > 0) {
+        _childManager.removeChild(firstChild);
+        leadingGarbage -= 1;
+      }
+      while (trailingGarbage > 0) {
+        _childManager.removeChild(lastChild);
+        trailingGarbage -= 1;
+      }
+    });
+  }
+
+  /// Returns the index of the given child, as given by the
+  /// [SliverMultiBoxAdaptorParentData.index] field of the child's [parentData].
+  int indexOf(RenderBox child) {
+    assert(child != null);
+    final SliverMultiBoxAdaptorParentData childParentData = child.parentData;
+    assert(childParentData.index != null);
+    return childParentData.index;
+  }
+
+  /// Returns the scroll offset of the given child, as given by the
+  /// [SliverMultiBoxAdaptorParentData.scrollOffset] field of the child's [parentData].
+  @protected
+  double offsetOf(RenderBox child) {
```

Comment:
offsetOf and indexOf should both either be protected or not; right now it's inconsistent